### PR TITLE
pynxtools_stm reader compatible with  and

### PR DESCRIPTION
**Pull request by RubelMozumder**
Modification for `scan_field` and `scan_offset`.

- [x] Fix the function: https://github.com/FAIRmat-NFDI/pynxtools-stm/blob/main/pynxtools_stm/sts_file_parser.py#L631
- [x] Check ci/cd that might contain working branch from pynxtools.

### CODE REVIEW

File: pynxtools_stm/reader.py [hunk 1]
```diff
@@ -347,12 +250,33 @@ def read(
...
+        #         filled_template[key] = val
+
         for key, val in template.items():
-            if val is not None:
+            if isinstance(val, np.ndarray):
                 filled_template[key] = val
+            elif val not in (None, ""):
+                try:
+                    # Handle numpy array, list, tuple, etc.
+                    if str(val) not in (
```
Comment (lukaspie) [inline pynxtools_stm/reader.py:266]: AFAICT, this is a repetition of the same (or similar) functionality in the `helpers.py/to_intended_t` function. Why not use that function here?

File: pynxtools_stm/stm_file_parser.py [hunk 2]
```diff
@@ -283,8 +269,15 @@ def fill_out_NXdata_group():
                         template[signal_attr] = data_field_nm.lower()
                     else:
                         template[auxiliary_signals_attr].append(data_field_nm.lower())
-                for axis, axis_data in zip(axes_name, axes_data):
+
+                if len(axes_data) != len(axes_units):
+                    missing = len(axes_data) - len(axes_units)
+                    axes_units.extend([""] * missing)
+                for axis, axis_data, unit in zip(axes_name, axes_data, axes_units):
+                    template[f"{nxdata_grp}/{axis}"] = axis_data
```
Comment (lukaspie) [inline pynxtools_stm/stm_file_parser.py:277]: Lines 277 and 278 are the same, but you are not using `axes_name`. I guess that's unintended.

File: pynxtools_stm/stm_file_parser.py [hunk 3]
```diff
@@ -317,18 +309,36 @@ def get_dimension_info(self, config_dict, data_dict):
             length on (x, y)-dimenstion and one last unknown values.
         """
         scanfield: str = ""
+        scan_range: str = ""
+        scan_offset: str = ""
+        scintific_num_pattern = r"[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?"
```
Comment (lukaspie) [inline pynxtools_stm/stm_file_parser.py:314]:
```suggestion
        scientific_num_pattern = r"[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?"
```

File: pynxtools_stm/stm_file_parser.py [hunk 4]
```diff
@@ -317,18 +309,36 @@ def get_dimension_info(self, config_dict, data_dict):
...
-                        "Scanfield data missing: /ENTRY[entry]/INSTRUMENT[instrument]"
-                        "/ENVIRONMENT[environment]/scan_control/positioner/scanfield"
-                    )
+            elif (
+                "/ENTRY[entry]/INSTRUMENT[instrument]/ENVIRONMENT[environment]"
+                "/scan_control/scan_range"
+            ) == key:
+                scan_range = data_dict.get(val, None)
+                if scan_range:
+                    scan_range = re.findall(scintific_num_pattern, scan_range)
```
Comment (lukaspie) [inline pynxtools_stm/stm_file_parser.py:328]: see above
```suggestion
                    scan_range = re.findall(scientific_num_pattern, scan_range)
```

File: pynxtools_stm/stm_file_parser.py [hunk 5]
```diff
@@ -317,18 +309,36 @@ def get_dimension_info(self, config_dict, data_dict):
...
+                    template[key] = to_intended_t(scan_range)
+            elif (
+                "/ENTRY[entry]/INSTRUMENT[instrument]/ENVIRONMENT[environment]"
+                "/scan_control/scan_offset"
+            ) == key:
+                scan_offset = data_dict.get(val, None)
+                if scan_offset:
+                    scan_offset = re.findall(scintific_num_pattern, scan_offset)
```
Comment (lukaspie) [inline pynxtools_stm/stm_file_parser.py:336]:
```suggestion
                    scan_offset = re.findall(scientific_num_pattern, scan_offset)
```

File: pynxtools_stm/stm_file_parser.py [hunk 6]
```diff
@@ -317,18 +309,36 @@ def get_dimension_info(self, config_dict, data_dict):
...
+                    template[key] = to_intended_t(scan_offset)
+        if (not scan_offset or not scan_range) and not scanfield:
+            raise KeyError(
+                "Scanfield, scan_range and scan_offset are not found in raw data file."
```
Comment (lukaspie) [inline pynxtools_stm/stm_file_parser.py:340]:
```suggestion
                "Scanfield, scan_range, and scan_offset are not found in raw data file."
```
Based on the if clause, the conditions is a bit different. It is either scanfield and scan_offset OR scanfield and scan_range are not found. But probably okay like this.

File: pynxtools_stm/stm_file_parser.py [hunk 7]
```diff
@@ -404,6 +419,80 @@ def from_sxm_file_into_template(self, template, config_dict, eln_data_dict):
         # The following function can be used later it link come true in application def.
         # link_implementation(template, nxdl_key_to_modified_key)
         link_seperation_from_hard_code(template, nxdl_key_to_modified_key)
+        self.set_default_values(template)
+
+    @staticmethod
+    def fill_temp_with_required_metadata(template, data_dict, config_dict):
```
Comment (lukaspie) [inline pynxtools_stm/stm_file_parser.py:425]: I am not sure why you have this method seperated from the rest of the metadata. Is it just to hardcode some stuff? In that case, it should be renamed.

**Review by lukaspie — COMMENTED** [review]
Some small comments, feel free to ignore if they are not appropriate.

File: pynxtools_stm/reader.py (hunk 1 above)
Comment (RubelMozumder) [reply pynxtools_stm/reader.py:266]: Removed!

File: pynxtools_stm/stm_file_parser.py (hunk 7 above)
Comment (RubelMozumder) [reply pynxtools_stm/stm_file_parser.py:425]: This methods independent of the object variables. But still bundled with STM file from Nanonis vendors.

File: pynxtools_stm/stm_file_parser.py (hunk 6 above)
Comment (RubelMozumder) [reply pynxtools_stm/stm_file_parser.py:340]: Noted!

File: pynxtools_stm/stm_file_parser.py (hunk 5 above)
Comment (RubelMozumder) [reply pynxtools_stm/stm_file_parser.py:336]: Noted!

File: pynxtools_stm/stm_file_parser.py (hunk 4 above)
Comment (RubelMozumder) [reply pynxtools_stm/stm_file_parser.py:328]: Noted!

File: pynxtools_stm/stm_file_parser.py (hunk 3 above)
Comment (RubelMozumder) [reply pynxtools_stm/stm_file_parser.py:314]: Noted!

File: pynxtools_stm/stm_file_parser.py (hunk 2 above)
Comment (RubelMozumder) [reply pynxtools_stm/stm_file_parser.py:277]: Noted!

**Pull request merged**